cpp/src/arrow/python/common.h

```diff
@@ -185,6 +185,62 @@ class ARROW_PYTHON_EXPORT OwnedRefNoGIL : public OwnedRef {
   }
 };
 
+template <typename Fn>
+struct BoundFunction;
+
+template <typename... Args>
+struct BoundFunction<void(PyObject*, Args...)> {
+  // We bind `cdef void fn(object, ...)` to get a `Status(...)`
+  // where the Status contains any Python error raised by `fn`
+  using Unbound = void(PyObject*, Args...);
+  using Bound = Status(Args...);
+
+  BoundFunction(Unbound* unbound, PyObject* bound_arg)
+      : bound_arg_(bound_arg), unbound_(unbound) {}
+
+  Status Invoke(Args... args) const {
+    unbound_(bound_arg_.obj(), std::forward<Args>(args)...);
+    return CheckPyError();
+  }
```

Comment:
If this is meant to be called from arbitrary C++ code, it should probably take the GIL before anything else. Perhaps reuse `SafeCallIntoPython`?

Reply:
SafeCallIntoPython is awkward in the presence of void return, I'll just use PyAcquireGIL